File: apps/api/src/modules/proxy/logger.ts
import type { Database } from "@raven/db";
import { requestLogs } from "@raven/db";
import type { Redis } from "ioredis";
import { publishEvent } from "@/lib/events";
import { incrementBudgetSpend } from "./budget-check";

export interface LogData {
  virtualKeyId: string;
  provider: string;
  providerConfigId: string;
  providerConfigName: string | null;
  model: string;
  method: string;
  path: string;
  statusCode: number;
  inputTokens: number;
  outputTokens: number;
  reasoningTokens: number;
  cost: number;
  latencyMs: number;
  cachedTokens: number;
  cacheHit: boolean;
  endUser: string | null;
  hasImages: boolean;
  imageCount: number;
  hasToolUse: boolean;
  toolCount: number;
  toolNames: readonly string[];
  sessionId: string | null;
  userAgent: string | null;
  guardrailMatches?: readonly {
    ruleName: string;
    ruleType: string;
    action: string;
    matchedContent: string;
  }[];
}

// --- Write buffer for batching log inserts ---
const FLUSH_INTERVAL = 2000;
const MAX_BUFFER = 100;

let bufferDb: Database | null = null;
const logBuffer: (typeof requestLogs.$inferInsert)[] = [];
let flushTimer: ReturnType<typeof setInterval> | null = null;

const startFlushTimer = (): void => {
  if (flushTimer) return;
  flushTimer = setInterval(() => {
    void flushLogBuffer();
  }, FLUSH_INTERVAL);
};

export const flushLogBuffer = async (): Promise<void> => {
  if (logBuffer.length === 0 || !bufferDb) return;

  const batch = logBuffer.splice(0, logBuffer.length);

  await bufferDb
    .insert(requestLogs)
    .values(batch)
    .catch((err) => console.error("Failed to flush log buffer:", err));
};

const bufferLogEntry = (db: Database, data: LogData): void => {
  bufferDb = db;
  logBuffer.push({
    cachedTokens: data.cachedTokens,
    cacheHit: data.cacheHit,
    cost: data.cost.toFixed(6),
    endUser: data.endUser,
    hasImages: data.hasImages,
    hasToolUse: data.hasToolUse,
    imageCount: data.imageCount,
    inputTokens: data.inputTokens,
    latencyMs: data.latencyMs,
    method: data.method,
    model: data.model,
    outputTokens: data.outputTokens,
    path: data.path,
    provider: data.provider,
    providerConfigId: data.providerConfigId,
    reasoningTokens: data.reasoningTokens,
    sessionId: data.sessionId,
    statusCode: data.statusCode,
    toolCount: data.toolCount,
    toolNames: data.toolNames.length > 0 ? [...data.toolNames] : undefined,
    userAgent: data.userAgent,
    virtualKeyId: data.virtualKeyId
  });
  startFlushTimer();

  if (logBuffer.length >= MAX_BUFFER) {
    void flushLogBuffer();
  }
};

export interface BudgetContext {
  redis: Redis;
}

export const logAndPublish = (
  db: Database,
  data: LogData,
  budgetCtx?: BudgetContext
): void => {
  void publishEvent("request.created", {
    ...data,
    cost: data.cost.toFixed(6),
    toolNames: data.toolNames.length > 0 ? [...data.toolNames] : []
  });

  bufferLogEntry(db, data);

  if (budgetCtx && data.cost > 0) {
    void incrementBudgetSpend(
      db,
      budgetCtx.redis,
      data.virtualKeyId,
      data.cost
    );
  }
};

File: apps/api-py/src/proxy/logger.py
import json
import logging
from decimal import Decimal

from sqlalchemy.ext.asyncio import AsyncSession

from src.models.request_log import RequestLog
from src.redis import get_redis

logger = logging.getLogger("raven.proxy.logger")


async def log_request(
    *,
    session: AsyncSession,
    virtual_key_id: str,
    provider: str,
    provider_config_id: str,
    model: str,
    method: str,
    path: str,
    status_code: int,
    input_tokens: int = 0,
    output_tokens: int = 0,
    cached_tokens: int = 0,
    reasoning_tokens: int = 0,
    cost: float = 0,
    latency_ms: int = 0,
    cache_hit: bool = False,
    session_id: str | None = None,
    end_user: str | None = None,
    user_agent: str | None = None,
    request_body: str | None = None,
    response_body: str | None = None,
) -> None:
    try:
        log_entry = RequestLog(
            virtual_key_id=virtual_key_id,
            provider=provider,
            provider_config_id=provider_config_id,
            model=model,
            method=method,
            path=path,
            status_code=status_code,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cached_tokens=cached_tokens,
            reasoning_tokens=reasoning_tokens,
            cost=Decimal(str(cost)),
            latency_ms=latency_ms,
            cache_hit=cache_hit,
            session_id=session_id,
            end_user=end_user,
            user_agent=user_agent,
            request_body=request_body,
            response_body=response_body,
        )
        session.add(log_entry)
        await session.commit()

        # Publish event for real-time dashboard
        redis = get_redis()
        event_data = {
            "provider": provider,
            "model": model,
            "statusCode": status_code,
            "inputTokens": input_tokens,
            "outputTokens": output_tokens,
            "cost": cost,
            "latencyMs": latency_ms,
            "cacheHit": cache_hit,
        }
        await redis.publish("raven:events", json.dumps({"type": "request.completed", "data": event_data}))

    except Exception:
        logger.exception("Failed to log request")
